cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go

```diff
@@ -0,0 +1,226 @@
+	data, ok := c.(controlPlanePrepareData)
+	if !ok {
+		return errors.New("certs phase invoked with an invalid data struct")
+	}
+
+	// Skip if this is not a control plane
+	if data.Cfg().ControlPlane == nil {
+		return nil
+	}
+
+	cfg, err := data.InitCfg()
+	if err != nil {
+		return err
+	}
+
+	// Generate missing certificates (if any)
+	return certsphase.CreatePKIAssets(cfg)
+}
+
+func runKubeconfigPhaseLocal(c workflow.RunData) error {
+	data, ok := c.(controlPlanePrepareData)
+	if !ok {
+		return errors.New("kubeconfig phase invoked with an invalid data struct")
+	}
+
+	// Skip if this is not a control plane
+	if data.Cfg().ControlPlane == nil {
+		return nil
+	}
+
+	cfg, err := data.InitCfg()
+	if err != nil {
+		return err
+	}
+
+	// Generate kubeconfig files for controller manager, scheduler and for the admin/kubeadm itself
+	// NB. The kubeconfig file for kubelet will be generated by the TLS bootstrap process in
+	// following steps of the join --experimental-control plane workflow
+	if err := kubeconfigphase.CreateJoinControlPlaneKubeConfigFiles(kubeadmconstants.KubernetesDir, cfg); err != nil {
+		return errors.Wrap(err, "error generating kubeconfig files")
+	}
+
+	return nil
+}
+
+func runControlPlaneJoinSubPhase(component string) func(c workflow.RunData) error {
+	return func(c workflow.RunData) error {
+		data, ok := c.(controlPlanePrepareData)
+		if !ok {
+			return errors.New("control-plane-prepare phase invoked with an invalid data struct")
+		}
+
+		// Skip if this is not a control plane
+		if data.Cfg().ControlPlane == nil {
+			return nil
+		}
+
+		cfg, err := data.InitCfg()
+		if err != nil {
+			return err
+		}
+
+		// Creates static pod manifests file for the control plane components to be deployed on this node
+		// Static pods will be created and managed by the kubelet as soon as it starts
+		fmt.Printf("[control-plane-prepare] Creating static Pod manifest for %q\n", component)
+		return controlplane.CreateStaticPodFiles(kubeadmconstants.GetStaticPodDirectory(), &cfg.ClusterConfiguration, &cfg.LocalAPIEndpoint, component)
+	}
+}
+
+func runCheckLocalEtcdStatus(c workflow.RunData) error {
+	data, ok := c.(controlPlanePrepareData)
+	if !ok {
+		return errors.New("check-local-etcd-status phase invoked with an invalid data struct")
+	}
+
+	// Skip if this is not a control plane
+	if data.Cfg().ControlPlane == nil {
+		return nil
+	}
+
+	cfg, err := data.InitCfg()
+	if err != nil {
+		return err
+	}
+
+	if cfg.Etcd.External != nil {
+		fmt.Println("[control-plane-prepare] Skipping etcd check in external mode")
+		return nil
+	}
+
+	// Checks that the etcd cluster is healthy
+	// NB. this check cannot be implemented before because it requires the admin.conf and all the certificates
+	//     for connecting to etcd already in place
+	kubeConfigFile := filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.AdminKubeConfigFileName)
+
+	client, err := kubeconfigutil.ClientSetFromFile(kubeConfigFile)
```

Comment:
Client should be part of JoinData and be implemented as a singleton

Reply:
I will refactor this.

Reply:
On a second pass we are creating three kind of Clientsets: admin, bootstrap kubelet and kubelet.

The operation of creating the clientset should be cheap (i.e. no remote actions), I wonder if having this as a singleton backed up by a map is making us gain something?

Reply:
Added a second commit (to be squashed when we are done with review). Is this what you mean?